src/index.ts

```diff
@@ -19,37 +19,55 @@ const run: Edkt['run'] = async (config) => {
     if (!hasConsent) return;
   }
 
-  // This is a no-op if undefined, equals current value or lesser than 0
+  // This is a no-op if undefined or lesser than 0
   viewStore.setStorageSize(featureStorageSize);
   viewStore.insert(pageFeatures, pageMetadata);
 
-  const matchedAudiences = audienceDefinitions
+  const currentTS = timeStampInSecs();
+
+  const newMatchedAudiences = audienceDefinitions
+    // drop prev matched audiences with same version
     .filter((audience) => {
-      return !matchedAudienceStore.matchedAudienceIds.includes(audience.id);
+      return !matchedAudienceStore.matchedAudiences.some(
+        ({ id, version }) => audience.id === id && version === audience.version
+      );
     })
+    // translate audience definitions into engine queries
     .map((audience) => {
       return {
         ...audience,
         conditions: engine.translate(audience),
       };
     })
-    .map((audience) => {
-      const currentTS = timeStampInSecs();
+    // check if query matches any pageViews
+    .map((query) => {
       const pageViewsWithinLookBack = viewStore.pageViews.filter((pageView) => {
-        return audience.lookBack === 0
+        return query.lookBack === 0
           ? true
-          : pageView.ts > currentTS - audience.lookBack;
+          : pageView.ts > currentTS - query.lookBack;
       });
       return {
-        id: audience.id,
+        id: query.id,
+        version: query.version,
         matchedAt: currentTS,
-        expiresAt: currentTS + audience.ttl,
+        expiresAt: currentTS + query.ttl,
         matchedOnCurrentPageView: true,
-        matched: engine.check(audience.conditions, pageViewsWithinLookBack),
+        matched: engine.check(query.conditions, pageViewsWithinLookBack),
       };
     })
+    // keep only matched audiences
     .filter((audience) => audience.matched);
 
+  // We also want to keep prev matched audiences with matching versions
+  const prevMatchedAudiences = matchedAudienceStore.matchedAudiences.filter(
+    (matched) =>
+      audienceDefinitions.some(
+        ({ id, version }) => matched.id === id && matched.version === version
+      )
+  );
```

Comment:
This step has to happen before the check to satisfy:

> if a user has matched version 1, and the most recent audience version available is 1, we should not run the checking.

Right?

Reply:
The skipping happens at the [first filter on the processing op](https://github.com/AirGrid/edgekit/pull/130/files/d626620d95f3147789e9a70ec8401a48c3fc051e#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R31-R33).
This op actually takes back the `pageViews` that were matched before and for which the `audienceDefinition` were not changed.

This division takes care of the deletion/version-bumping automatically.
Another option is to skip all the pageViews with `<= || ===` version to their `audienceDefinition`, bump the ones that match the new definition and then delete manually the ones which would not get bumped.

I found the chosen algorithm simpler. It also makes the `setMatchedAudiences` actually a setter.
It was appending before the change.

Reply:
> The skipping happens at the first filter on the processing op.

yeah sorry missed this! this part looks good.

> This op actually takes back the pageViews that were matched before and for which the audienceDefinition were not changed.

Sorry this does not mates sense to me, why do you "take back pageviews"? 

> This division takes care of the deletion/version-bumping automatically.

Where is this deletion? Sorry I am unclear here again.

> Another option is to skip all the pageViews with <= || === version to their audienceDefinition, bump the ones that match the new definition and then delete manually the ones which would not get bumped.

The pageview feature version is not related to the audience definition global version - they should not be compared.

Reply:
Sorry, I was thinking `matchedAudience` and writing `pageView`.
So, let's try again!

The changes outline are as follows:

- First, now the `setMatchedAudiences` from `MatchedAudiencesStore` is actually a setter, so it will not append anymore, but actually overwrite the `matchedAudiences` and `matchedAudiencesId` props.

With that in mind, for the processing:
- First, for each `matchedAudience` we skip the ones which still shares `version` with `audienceDefinition` (the prev matched and not version bumped ones);
- Then we do all the processing, looking through `pageViews` for new matches. Once we've finished, on `newMatchedAudiences` var we have all the matched audiences. Those which were not bumped but was not previously matched and those with a bumped version which matches again;
- Now we have all the newly matched audiences and the old with a bumped version that matches again, so we need to recover the ones that have matched previously, and of which definition does not have a bumped version. These are exactly the ones which we filtered at the start. They are captured over the `prevMatchedAudiences` var;
- Merging these vars, we have now all the interesting audiences, with correct versions. We can now set it as the `matchetAudiences` on `MatchedAudiencesStore`, overwriting the prev array. The overwriting takes care of deleting the old ones with an outdated version.

Does it make sense to you?

Reply:
Yes this does make sense.